include/revng/Pipeline/ContainerSet.h

```diff
@@ -0,0 +1,143 @@
+#pragma once
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include <cstddef>
+#include <fstream>
+#include <memory>
+#include <optional>
+#include <set>
+#include <sstream>
+#include <string>
+
+#include "llvm/ADT/StringMap.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/IR/Module.h"
+#include "llvm/Support/Casting.h"
+#include "llvm/Support/Error.h"
+#include "llvm/Support/FileSystem.h"
+#include "llvm/Support/MemoryBuffer.h"
+
+#include "revng/Pipeline/Container.h"
+#include "revng/Pipeline/ContainerFactory.h"
+#include "revng/Pipeline/Contract.h"
+#include "revng/Pipeline/Target.h"
+#include "revng/Support/Assert.h"
+
+namespace Pipeline {
+
+/// A map from containerName to a MaybeContainer, which is initially empty.
+///
+/// This class is used mostly in two points, as the backing containers
+/// of each step, as well as the containers containing the target extracted
+/// from each step to perform the computation.
+class ContainerSet {
+private:
+  using Map = llvm::StringMap<std::unique_ptr<ContainerBase>>;
+
+public:
+  using const_iterator = Map::const_iterator;
+  using iterator = Map::iterator;
+  using value_type = Map::value_type;
+
+private:
+  Map Content;
+  llvm::StringMap<const ContainerFactory *> Factories;
+
+public:
+  ContainerSet(Map Content) : Content(std::move(Content)) {}
+  ContainerSet() = default;
+  ContainerSet(ContainerSet &&) = default;
+  ContainerSet &operator=(ContainerSet &&) = default;
+  ~ContainerSet() = default;
+  ContainerSet(const ContainerSet &) = delete;
+  ContainerSet &operator=(const ContainerSet &) = delete;
+
+  ContainerSet cloneFiltered(const ContainersStatus &Targets);
+
+  void mergeBack(ContainerSet &&Other) {
+    for (auto &Entry : Other.Content) {
+      revng_assert(containsOrCanCreate(Entry.first()));
+
+      auto &LContainer = Content.find(Entry.first())->second;
+      auto &RContainer = Entry.second;
+      if (RContainer == nullptr)
+        continue;
+
+      if (LContainer == nullptr)
+        LContainer = std::move(RContainer);
+      else
+        LContainer->mergeBack(std::move(*RContainer));
+    }
+  }
+
+  ContainerBase &operator[](llvm::StringRef Name) {
+    revng_assert(containsOrCanCreate(Name));
+    if (Content[Name] == nullptr)
+      Content[Name] = (*Factories[Name]) (Name);
+    auto &Pointer = Content.find(Name)->second;
+    revng_assert(Pointer != nullptr);
+    return *Pointer;
+  }
+
+  ContainerBase &at(llvm::StringRef Name) {
+    revng_assert(contains(Name));
+    return *Content.find(Name)->second;
+  }
```

Comment:
Do you see a legitimate use case for the user to be interested in distinguishing whether it's there or it needs to be created on the fly?

We might as well simplify the API and prevent misuses by just exposing `operator[]`.

Reply:
i tried removing but there are point in which you do want actually look into the without materializing them, such as when one want to store them onto disk, if they are not materialized, they should not be stored.

I guess i can drop them, but it is more correct to use contains followed by at rather than operator[] since if the check is incurrectly removed the at will assert and operator[] will silently break.

Reply:
i am thinking really hard about this one, the issue is that since it's lazy it does not matter what we do, it will always have a semantics for operator[] different from the one of a map.

I can have the indended behaviour if i mark the lazy members as mutable, but that is ugly as well.